### PR TITLE
Remove segment group from delta writer

### PR DESCRIPTION
**Pull request by kangpinghuang**
1. remove SegmengGroup and ColumnDataWriter from DeltaWriter, replace them with Rowset and RowsetBuilder
2. Add RowsetBuilderContext and RowsetBuilderContextBuilder
3. Optimize RowsetBuilder to support pending Rowset and non-pending Rowset

### CODE REVIEW

File: be/src/olap/delta_writer.cpp [hunk 1]
```diff
@@ -143,20 +148,7 @@ OLAPStatus DeltaWriter::write(Tuple* tuple) {
...
-        _cur_segment_group = nullptr; 
-        DCHECK(_cur_segment_group != nullptr) << "failed to malloc SegmentGroup";
-        _cur_segment_group->acquire();
-        _cur_segment_group->set_load_id(_req.load_id);
-        _segment_group_vec.push_back(_cur_segment_group);
-
-        SAFE_DELETE(_writer);
-        _writer = ColumnDataWriter::create(_cur_segment_group, true, _tablet->compress_kind(), _tablet->bloom_filter_fpp());
-        DCHECK(_writer != nullptr) << "memory error occur when creating writer";
+        RETURN_NOT_OK(_mem_table->flush(_rowset_builder.get()));
```
Comment (chaoyli) [inline be/src/olap/delta_writer.cpp]: May use shared_ptr will OK?

File: be/src/olap/delta_writer.cpp [hunk 2]
```diff
@@ -172,15 +164,18 @@ OLAPStatus DeltaWriter::close(google::protobuf::RepeatedPtrField<PTabletInfo>* t
...
+    RETURN_NOT_OK(_mem_table->close(_rowset_builder.get()));
 
     OLAPStatus res = OLAP_SUCCESS;
     //add pending data to tablet
-    RETURN_NOT_OK(_tablet->add_pending_version(_req.partition_id, _req.transaction_id, nullptr));
-    for (SegmentGroup* segment_group : _segment_group_vec) {
-        RETURN_NOT_OK(_tablet->add_pending_segment_group(segment_group));
-        RETURN_NOT_OK(segment_group->load());
-    }
+    RETURN_NOT_OK(_tablet->add_pending_version(_req.partition_id, _req.txn_id, nullptr));
```
Comment (chaoyli) [inline be/src/olap/delta_writer.cpp]: tablet has no add_pending_version interface, you should add it into RowsetMetaManger.

File: be/src/olap/delta_writer.cpp [hunk 3]
```diff
@@ -172,15 +164,18 @@ OLAPStatus DeltaWriter::close(google::protobuf::RepeatedPtrField<PTabletInfo>* t
...
+    // use rowset meta manager to save meta
+    _cur_rowset = _rowset_builder->build();
+    RowsetMetaManager::save(
+            _tablet->data_dir()->get_meta(),
+            _cur_rowset->rowset_id(),
+            _cur_rowset->rowset_meta().get());
```
Comment (chaoyli) [inline be/src/olap/delta_writer.cpp]: Also, use std::shared_ptr<RowsetMeta> instead

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: be/src/olap/delta_writer.cpp (hunk 1 above)
Comment (kangpinghuang) [reply be/src/olap/delta_writer.cpp]: OK, I will modify it

File: be/src/olap/delta_writer.cpp (hunk 2 above)
Comment (kangpinghuang) [reply be/src/olap/delta_writer.cpp]: OK, I think this should be removed

File: be/src/olap/delta_writer.cpp (hunk 3 above)
Comment (kangpinghuang) [reply be/src/olap/delta_writer.cpp]: OK, I will modify it

File: be/src/olap/delta_writer.cpp [hunk 4]
```diff
@@ -29,34 +31,27 @@ OLAPStatus DeltaWriter::open(WriteRequest* req, DeltaWriter** writer) {
 
 DeltaWriter::DeltaWriter(WriteRequest* req)
     : _req(*req), _tablet(nullptr),
-      _cur_segment_group(nullptr), _new_tablet(nullptr),
-      _writer(nullptr), _mem_table(nullptr),
+      _cur_rowset(nullptr), _related_rowset(nullptr), _new_tablet(nullptr),
```
Comment (chaoyli) [inline be/src/olap/delta_writer.cpp]: The two variable names  _related_rowset and _new_tablet should be consistent.

File: be/src/olap/rowset/rowset_builder_context.h [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    // properties for non-pending rowset
+    Version version;
+    VersionHash version_hash;
+
+    // properties for pending rowset
+    int64_t txn_id;
+    PUniqueId load_id;
+};
+
+class RowsetBuilderContextBuilder {
```
Comment (chaoyli) [inline be/src/olap/rowset/rowset_builder_context.h]: This name may be thought deeply. It may confused the others.

File: be/src/olap/memtable.h [hunk 6]
```diff
@@ -23,6 +23,7 @@
 #include "olap/schema.h"
 #include "olap/skiplist.h"
 #include "runtime/tuple.h"
+#include "olap/rowset/rowset_builder.h"
 
 namespace doris {
```
Comment (chaoyli) [inline be/src/olap/memtable.h]: ColumnDataWriter declaration should be removed.

File: be/src/olap/rowset/rowset_builder_context.h [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        _rowset_builder_context.partition_id = partition_id;
+        return *this;
+    }
+
+    RowsetBuilderContextBuilder& set_rowset_type(RowsetTypePB rowset_type) {
+        _rowset_builder_context.rowset_type = rowset_type;
+        return *this;
+    }
+
+    RowsetBuilderContextBuilder& set_rowset_path_prefix(std::string rowset_path_prefix) {
```
Comment (chaoyli) [inline be/src/olap/rowset/rowset_builder_context.h]: const string&

File: be/src/olap/rowset/rowset_builder_context.h [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        _rowset_builder_context.tablet_schema_hash = tablet_schema_hash;
+        return *this;
+    }
+    
+    RowsetBuilderContextBuilder& set_partition_id(int64_t partition_id) {
+        _rowset_builder_context.partition_id = partition_id;
+        return *this;
+    }
+
+    RowsetBuilderContextBuilder& set_rowset_type(RowsetTypePB rowset_type) {
```
Comment (chaoyli) [inline be/src/olap/rowset/rowset_builder_context.h]: const RowsetTypePB&

File: be/src/olap/rowset/rowset_builder_context.h [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        _rowset_builder_context.compress_kind = compress_kind;
+        return *this;
+    }
+
+    RowsetBuilderContextBuilder& set_bloom_filter_fpp(double bloom_filter_fpp) {
+        _rowset_builder_context.bloom_filter_fpp = bloom_filter_fpp;
+        return *this;
+    }
+
+    RowsetBuilderContextBuilder& set_rowset_state(RowsetStatePB rowset_state) {
```
Comment (chaoyli) [inline be/src/olap/rowset/rowset_builder_context.h]: const RowsetStatePB&

File: be/src/olap/rowset/rowset_builder_context.h [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+        _rowset_builder_context.num_short_key_fields = num_short_key_fields;
+        return *this;
+    }
+
+    RowsetBuilderContextBuilder& set_num_rows_per_row_block(size_t num_rows_per_row_block) {
+        _rowset_builder_context.num_rows_per_row_block = num_rows_per_row_block;
+        return *this;
+    }
+
+    RowsetBuilderContextBuilder& set_compress_kind(CompressKind compress_kind) {
```
Comment (chaoyli) [inline be/src/olap/rowset/rowset_builder_context.h]: const CompressKind&

File: be/src/olap/rowset/alpha_rowset_builder.cpp [hunk 11]
```diff
@@ -72,27 +83,47 @@ OLAPStatus AlphaRowsetBuilder::flush() {
...
+            const std::vector<KeyRange>* column_statistics = &(segment_group->get_column_statistics());
+            if (column_statistics != nullptr) {
+                for (size_t i = 0; i < column_statistics->size(); ++i) {
+                    ColumnPruning* column_pruning = segment_group_pb.add_column_pruning();
+                    column_pruning->set_min(column_statistics->at(i).first->to_string());
+                    column_pruning->set_max(column_statistics->at(i).second->to_string());
+                    column_pruning->set_null_flag(column_statistics->at(i).first->is_null());
+                }
             }
+            segment_group_pb.set_empty(segment_group->empty());
```
Comment (chaoyli) [inline be/src/olap/rowset/alpha_rowset_builder.cpp]: std::reinterpret_cast<> is better to convert from base class to inherited class.

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: be/src/olap/rowset/rowset_builder_context.h (hunk 8 above)
Comment (kangpinghuang) [reply be/src/olap/rowset/rowset_builder_context.h]: This is a enum, not a object

File: be/src/olap/rowset/rowset_builder_context.h (hunk 9 above)
Comment (kangpinghuang) [reply be/src/olap/rowset/rowset_builder_context.h]: This is a enum too.

File: be/src/olap/rowset/rowset_builder_context.h (hunk 10 above)
Comment (kangpinghuang) [reply be/src/olap/rowset/rowset_builder_context.h]: this is a enum

File: be/src/olap/rowset/alpha_rowset_builder.cpp (hunk 11 above)
Comment (kangpinghuang) [reply be/src/olap/rowset/alpha_rowset_builder.cpp]: OK， I will fix it

**Pull request merged**